src/hammer-vlsi/hammer_vlsi/submit_command.py

```diff
@@ -0,0 +1,236 @@
+        else:
+            prog_args = remaining_args[0:ARG_DISPLAY_LEN-1] + "..."
+        prog_tag = prog_name + " " + prog_args
+
+        logger.debug("Executing subprocess: " + ' '.join(args))
+        subprocess_logger = logger.context("Exec " + prog_tag)
+        proc = subprocess.Popen(args, shell=False, stderr=subprocess.STDOUT, stdout=subprocess.PIPE, env=env, cwd=cwd)
+        atexit.register(proc.kill)
+
+        output_buf = ""
+        # Log output and also capture output at the same time.
+        while True:
+            line = proc.stdout.readline().decode("utf-8")
+            if line != '':
+                subprocess_logger.debug(line.rstrip())
+                output_buf += line
+            else:
+                break
+        # TODO: check errors
+
+        return output_buf
+
+    def read_settings(self, settings: Dict[str, Any], tool_namespace: str) -> None:
+        assert("Should never get here; local submission command does not have settings")
+
+class HammerLSFSubmitCommand(HammerSubmitCommand):
+
+    # TODO list:
+    #  - we need to log the command output
+
+    @property
+    def bsub_binary(self) -> str:
+        """ Get the LSF bsub binary location """
+        try:
+            return self._bsub_binary
+        except AttributeError:
+            raise ValueError("LSF bsub binary location not set")
+
+    @bsub_binary.setter
+    def bsub_binary(self, value: str) -> None:
+        """ Set the LSF bsub binary location """
+        if value == "" or value is None:
+            raise ValueError("bsub_binary cannot be empty or null")
+        self._bsub_binary = value
+
+    @property
+    def num_cpus(self) -> Optional[int]:
+        """ Get the number of CPUs to use """
+        try:
+            return self._num_cpus
+        except AttributeError:
+            raise ValueError("Did not set the number of CPUs to use (can be None)")
+
+    @num_cpus.setter
+    def num_cpus(self, value: int) -> None:
+        """ Set the number of CPUs to use """
+        self._num_cpus = value # type: Optional[int]
+
+    @property
+    def queue(self) -> Optional[str]:
+        """ Get the LSF queue to use """
+        try:
+            return self._queue
+        except AttributeError:
+            raise ValueError("Did not set the queue to use (can be None)")
+
+    @queue.setter
+    def queue(self, value: str) -> None:
+        """ Set the LSF queue to use """
+        self._queue = value # type: Optional[str]
+
+    @property
+    def extra_args(self) -> List[str]:
+        """ Get the extra LSF args to use """
+        try:
+            return self._extra_args
+        except AttributeError:
+            raise ValueError("Did not set the extra_args to use (can be [])")
+
+    @extra_args.setter
+    def extra_args(self, value: List[str]) -> None:
+        """ Set the extra LSF args to use """
+        self._extra_args = value
+
+    def read_settings(self, settings: Dict[str, Any], tool_name):
```

Comment:
Can the settings live in a separate `NamedTuple`?

Reply:
What would that buy us?

Reply:
This is done inb 499273